waiter/src/waiter/token.clj

```diff
@@ -516,48 +516,62 @@
+        (if editor?
+          (let [existing-token-parameters (sd/token->token-parameters kv-store token :include-deleted false)]
+            (doseq [parameter-name ["editor" "owner" "run-as-user"]]
+              (let [existing-value (get existing-token-parameters parameter-name)
+                    new-value (get new-token-data parameter-name)]
+                (when (not= existing-value new-value)
+                  (throw (ex-info (str "Not allowed to edit parameter " parameter-name)
+                                  {:authenticated-user authenticated-user
+                                   :existing-token-description existing-token-parameters
+                                   :parameter parameter-name
+                                   :parameter-exiting-value existing-value
+                                   :parameter-new-value new-value
+                                   :privileges {:editor? editor? :owner? owner?}
+                                   :status http-403-forbidden
+                                   :log-level :warn}))))))
+          ;; only check run-as-user rules when not running as editor
+          (when (and run-as-user (not= "*" run-as-user))
+            (when-not (authz/run-as? entitlement-manager authenticated-user run-as-user)
+              (throw (ex-info "Cannot run as user"
                               {:authenticated-user authenticated-user
-                               :owner owner
+                               :privileges {:editor? editor? :owner? owner?}
+                               :run-as-user run-as-user
                                :status http-403-forbidden
                                :log-level :warn})))))
-        (when (contains? new-token-metadata "last-update-time")
-          (throw (ex-info "Cannot modify last-update-time token metadata"
-                          {:status http-400-bad-request
-                           :token-metadata new-token-metadata
-                           :log-level :warn})))
-        (when (contains? new-token-metadata "last-update-user")
-          (throw (ex-info "Cannot modify last-update-user token metadata"
-                          {:status http-400-bad-request
-                           :token-metadata new-token-metadata
-                           :log-level :warn})))
-        (when (contains? new-token-metadata "root")
-          (throw (ex-info "Cannot modify root token metadata"
-                          {:status http-400-bad-request
-                           :token-metadata new-token-metadata
-                           :log-level :warn})))
-        (when (contains? new-token-metadata "previous")
-          (throw (ex-info "Cannot modify previous token metadata"
-                          {:status http-400-bad-request
-                           :token-metadata new-token-metadata
-                           :log-level :warn}))))
+        (if-not (str/blank? existing-owner)
```

Comment:
Isn't your claim that `existing-owner` is always available?

Reply:
If this check for existing owner is really checking whether the token exists or not, can we make that a bit more explicit?